File: examples/mcp-redact/python/wrap_tool_call.py
"""Server-side and client-side wrappers built on ``redact_tool_call.py``.

Both wrap a plain async callable matching the official Python MCP SDK's
structural shapes, verified while resolving issue #327 against ``mcp==2.2.0``
(``mcp.server.lowlevel.Server``'s ``on_call_tool`` constructor callback,
``mcp/server/lowlevel/server.py``: ``Callable[[ctx, CallToolRequestParams],
Awaitable[CallToolResult]]``; ``mcp.client.session.ClientSession.call_tool``,
``mcp/client/session.py``: ``(name, arguments) -> CallToolResult``) without
importing ``mcp`` itself -- ``CallToolResult`` is a plain dict tree here, the
same duck-typed choice
``examples/tracing-masking/python/redact_span_attributes.py`` makes for
OpenTelemetry's ``SpanProcessor``.

Real wiring:

.. code-block:: python

    import redact_secret
    from mcp.server.lowlevel import Server
    from wrap_tool_call import wrap_server_tool_handler

    async def read_file(ctx, params):
        ...  # returns a CallToolResult-shaped dict

    server = Server(
        "example",
        on_call_tool=wrap_server_tool_handler(
            read_file, redact_secret.scan_and_redact, on_finding=audit,
        ),
    )

.. code-block:: python

    import redact_secret
    from wrap_tool_call import wrap_client_call_tool

    call_tool = wrap_client_call_tool(session.call_tool, redact_secret.scan_and_redact, on_finding=audit)
    result = await call_tool("read_file", {"path": "/tmp/x"})
"""

from __future__ import annotations

from typing import Any, Callable, Optional

from redact_tool_call import build_blocked_result, redact_arguments, redact_tool_result

__all__ = ["emit_findings", "wrap_server_tool_handler", "wrap_client_call_tool"]


def _get_arguments(params: Any) -> dict[str, Any]:
    """Reads `.arguments` from either a plain dict (used by this module's
    own tests) or a pydantic `CallToolRequestParams` (the real SDK)."""
    if isinstance(params, dict):
        return params.get("arguments") or {}
    return getattr(params, "arguments", None) or {}


def _with_arguments(params: Any, arguments: dict[str, Any]) -> Any:
    if isinstance(params, dict):
        return {**params, "arguments": arguments}
    return params.model_copy(update={"arguments": arguments})


def _as_dict(value: Any) -> dict[str, Any]:
    """Normalizes either a plain dict or a pydantic `CallToolResult` to a
    plain dict tree, the shape `redact_tool_result` operates on."""
    if isinstance(value, dict):
        return value
    return value.model_dump(by_alias=True)


def emit_findings(findings: list[Any], on_finding: Optional[Callable[..., None]], scope: str) -> None:
    """Reported for every finding, including on a blocked outcome, as
    exactly the safe metadata ``scan_and_redact`` already returns -- never
    the input or a matched value. A raising ``on_finding`` is swallowed and
    never influences the redaction outcome. Not underscore-prefixed:
    ``agent_context.py`` imports it rather than keeping a second copy in
    this same directory."""
    if on_finding is None:
        return
    for finding in findings:
        try:
            on_finding(finding, scope=scope)
        except Exception:
            # The audit callback is best-effort: it never influences the
            # redaction outcome, so a raising callback is swallowed rather
            # than failing (or, worse, un-blocking) the call.
            pass


def wrap_server_tool_handler(
    handler: Callable[..., Any],
    scan_and_redact: Callable[..., Any],
    *,
    policy: Optional[Any] = None,
    limits: Optional[dict[str, int]] = None,
    redact_arguments_before_forwarding: bool = False,
    on_finding: Optional[Callable[..., None]] = None,
):
    """Wraps a server-side ``on_call_tool``-shaped handler,
    ``async def handler(ctx, params) -> CallToolResult`` where
    ``params.arguments`` is the tool's argument dict.

    Arguments are redacted first -- and, on a ``block`` finding there, the
    handler is never called, so a blocked secret never reaches the tool
    implementation -- then the handler's result is redacted before it is
    returned to the client. ``redact_arguments_before_forwarding`` defaults
    to ``False``: most tools need the real argument value to function (an
    API key they must actually send), so argument redaction is opt-in per
    tool, as the issue's false-positive/negative section documents.
    """
    if not callable(handler):
        raise TypeError("wrap_server_tool_handler: handler must be callable")

    async def redacting_tool_handler(ctx: Any, params: Any) -> Any:
        effective_params = params
        if redact_arguments_before_forwarding:
            args = _get_arguments(params)
            arg_outcome = redact_arguments(scan_and_redact, args, policy=policy, limits=limits)
            emit_findings(arg_outcome["findings"], on_finding, "argument")
            if arg_outcome["outcome"] == "blocked":
                return build_blocked_result()
            effective_params = _with_arguments(params, arg_outcome["arguments"])

        raw = await handler(ctx, effective_params)
        result_outcome = redact_tool_result(scan_and_redact, _as_dict(raw), policy=policy, limits=limits)
        emit_findings(result_outcome["findings"], on_finding, "result")
        if result_outcome["outcome"] == "blocked":
            return build_blocked_result()
        return result_outcome["result"]

    return redacting_tool_handler


def wrap_client_call_tool(
    call_tool: Callable[..., Any],
    scan_and_redact: Callable[..., Any],
    *,
    policy: Optional[Any] = None,
    limits: Optional[dict[str, int]] = None,
    on_finding: Optional[Callable[..., None]] = None,
):
    """Wraps a client-side ``call_tool``. Only the result is redacted --
    arguments are already being sent to the server at this point -- before
    it enters model context, matching the issue's client-side scope.
    """
    if not callable(call_tool):
        raise TypeError("wrap_client_call_tool: call_tool must be callable")

    async def redacting_call_tool(*args: Any, **kwargs: Any) -> Any:
        raw = await call_tool(*args, **kwargs)
        result_outcome = redact_tool_result(scan_and_redact, _as_dict(raw), policy=policy, limits=limits)
        emit_findings(result_outcome["findings"], on_finding, "result")
        if result_outcome["outcome"] == "blocked":
            return build_blocked_result()
        return result_outcome["result"]

    return redacting_call_tool

File: examples/mcp-redact/wrap-tool-call.mjs
/**
 * Server-side and client-side wrappers over one AI-context boundary, through
 * `@redact-secret/adapter-mcp` (the MCP boundary contract,
 * `docs/reference/mcp-boundary.md`, #612).
 *
 * Both wrap a plain function matching an MCP TypeScript SDK's structural
 * shapes, without importing an SDK. The adapter is tested with real SDK
 * instances of both supported lines (`@modelcontextprotocol/sdk`
 * `>=1.13.0 <=1.30.1`, `@modelcontextprotocol/client` and `/server`
 * `>=2.0.0 <=2.1.0`) over stdio and Streamable HTTP.
 *
 * Real wiring (`EXAMPLE_LIMITS` is in `agent-context.mjs`; limits are
 * mandatory and explicit):
 *
 * ```js
 * import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
 * import { createAiContextBoundary } from "@redact-secret/adapter-ai-context";
 * import { EXAMPLE_LIMITS } from "./agent-context.mjs";
 * import { wrapServerToolHandler } from "./wrap-tool-call.mjs";
 *
 * const boundary = await createAiContextBoundary({ ...EXAMPLE_LIMITS, onFinding: audit });
 * const server = new McpServer({ name: "example", version: "1.0.0" });
 * server.registerTool(
 *   "read_file",
 *   { inputSchema: { path: z.string() } },
 *   wrapServerToolHandler(readFileHandler, boundary),
 * );
 * ```
 *
 * ```js
 * import { Client } from "@modelcontextprotocol/sdk/client/index.js";
 * import { createAiContextBoundary } from "@redact-secret/adapter-ai-context";
 * import { EXAMPLE_LIMITS } from "./agent-context.mjs";
 * import { wrapClientCallTool } from "./wrap-tool-call.mjs";
 *
 * const boundary = await createAiContextBoundary({ ...EXAMPLE_LIMITS, onFinding: audit });
 * const callTool = wrapClientCallTool((params, { signal }) => client.callTool(params, undefined, { signal }), boundary);
 * const result = await callTool({ name: "read_file", arguments: { path: "/tmp/x" } });
 * ```
 *
 * The client side is the authoritative placement: the host applies the
 * boundary to every result it receives, including `isError: true` results,
 * before logging, persistence, or model context. A server-side wrapper is
 * preventive.
 *
 * Audit telemetry is the boundary's `onFinding(finding, { boundary })`,
 * called once per finding with the allowlisted safe metadata, labelled
 * `tool-result` or `tool-arguments`. A throwing callback is swallowed and
 * never changes an outcome.
 */

import { toCallToolResult } from "@redact-secret/adapter-mcp";

import { mcpBoundaryFor } from "./redact-tool-call.mjs";

function requireFunction(value, caller, name) {
  if (typeof value !== "function") throw new TypeError(`${caller}: ${name} must be a function`);
}

/**
 * Wraps a server-side tool handler with the adapter's `wrapToolHandler`.
 * With `redactArguments`, arguments are sanitized first and, on any non-`ok`
 * outcome, the handler is never called. The handler's result is sanitized
 * before it is returned. A handler that throws becomes the fixed tool-error
 * result, so the SDK never turns its `error.message` into result text.
 * Both steps use the request's signal (`extra.signal` on SDK 1.x,
 * `ctx.mcpReq.signal` on 2.x). `redactArguments` defaults to `false`: most
 * tools need the real argument value to function (an API key they must
 * actually send), so argument redaction is opt-in per tool.
 *
 * @param {(args: object, extra: unknown) => unknown | Promise<unknown>} handler
 * @param {import("@redact-secret/adapter-ai-context").AiContextBoundary} boundary
 * @param {{ redactArguments?: boolean, binaryContent?: "block" | "pass" }} [options]
 */
export function wrapServerToolHandler(handler, boundary, options = {}) {
  requireFunction(handler, "wrapServerToolHandler", "handler");
  return mcpBoundaryFor(boundary, { binaryContent: options.binaryContent }).wrapToolHandler(handler, {
    sanitizeArguments: options.redactArguments === true,
  });
}

/**
 * Wraps a client-side `callTool` with the adapter's `sanitizeToolCall`. The
 * result is sanitized before it enters model context. A rejected call (an
 * `McpError` or a transport error) becomes the fixed tool-error result, its
 * message never read. A cancelled call resolves to `null`: nothing is
 * delivered.
 *
 * @param {(params: unknown, options: { signal?: AbortSignal }) => unknown | Promise<unknown>} callTool
 * @param {import("@redact-secret/adapter-ai-context").AiContextBoundary} boundary
 * @param {{ binaryContent?: "block" | "pass" }} [options]
 */
export function wrapClientCallTool(callTool, boundary, options = {}) {
  requireFunction(callTool, "wrapClientCallTool", "callTool");
  const mcp = mcpBoundaryFor(boundary, { binaryContent: options.binaryContent });
  return async function redactingCallTool(params, { signal } = {}) {
    const outcome = await mcp.sanitizeToolCall((context) => callTool(params, { signal: context.signal }), { signal });
    return toCallToolResult(outcome);
  };
}
